specs/codecs/dag-cosmos/tendermint_chain.md

```diff
@@ -0,0 +1,355 @@
+type Block struct {
+	Header Header
+	Data Txs
+	Evidence EvidenceList
+	LastCommit Commit
+}
+
+# BlockID contains two distinct Merkle roots of the block.
+type BlockID struct {
+	Hash          HeaderCID
+	PartSetHeader PartSetHeader
+}
+
+# PartSetHeader is used for secure gossiping of the block during consensus
+# It contains the Merkle root of the complete serialized block cut into parts (ie. MerkleRoot(MakeParts(block))).
+type PartSetHeader struct {
+	Total Uint
+	Hash  BlockCID
+}
+
+# PartSet is the complete set of parts for a header
+type PartSet [Part]
+
+# Part is a section of bytes of a complete protobuf serialized block
+type Part struct {
+	Index Uint
+	Bytes HexBytes
+	Proof Proof
+}
+```
+
+## Light Block
+LightBlock is the core data structure of the light client.
+It combines two data structures needed for verification (SignedHeader & ValidatorSet).
+CID links to LightBlock use the SHA256 multihash and the LightBlock codec (tbd).
+```ipldsch
+# LightBlock is a SignedHeader and a ValidatorSet.
+# It is the basis of the light client
+type LightBlock struct {
+	SignedHeader SignedHeader
+	ValidatorSet  ValidatorSet
+}
+
+# SignedHeader is a header along with the commits that prove it.
+type SignedHeader struct {
+	Header Header
+	Commit Commit
+}
+```
+
+## Tendermint Merkle Node
+A Merkle tree node is a single node in a Merkle tree. A Merkle node is referenced by its hash.
+The top-level node in the tree is the root node, its hash is the Merkle root. Tendermint uses
+the RFC 6962 specification of a merkle tree, with sha256 as the hash function.
+* The hash of an inner node is `SHA_256(0x01 || left_hash || right_hash)`.
+* The hash of a leaf node is `SHA_256(0x00 || value)`.
+* The multicodec type used for a Merkle tree will depend on the content it stores
+```ipldsch
+type MerkleTreeNode union {
+    | MerkleTreeRootNode "root"
+    | MerkleTreeInnerNode "inner"
+    | MerkleTreeLeafNode "leaf"
+} representation keyed
+
+# MerkleTreeRootNode is the top-most node in a merkle tree; the root node of the tree.
+# It can be a leaf node if there is only one value in the tree
+# We explicitly type the root node for the purposes of handling Block and Header marshalling
+# since entire sub-dags need to be collapsed down and encoded into the protobuf object we need to be
+# able to check that the ipld.Node provided to the marshaller is a root node
+type MerkleTreeRootNode MerkleTreeNode
+
+# MerkleTreeInnerNode nodes contain two byte arrays which contain the hashes which link its two child nodes.
+type MerkleTreeInnerNode struct {
+    Left nullable &MerkleTreeNode
+    Right nullable &MerkleTreeNode
+}
+
+# Value union type used to handle the different values stored in leaf nodes in the different merkle trees
+type Value union {
+    | SimpleValidator "validator"
+    | Evidence "evidence"
+    | TxCID "tx"
+    | Part "part"
+    | ResponseDeliverTx "result"
+    | Bytes "header"
+    | CommitSig "commit"
+} representation keyed
```

Comment:
Are these keyed unions in the data itself, or is it a mechanism to produce usable data structures? I note the same use of `&MerkleTreeNode` throughout but then dealing with the differentiation at the `keyed` representation level of `Value`, but do these keys actually exist? Is the alternative of having separate `SimpleValidatorMerkleTreeNode` and friends just too verbose for all of these different types.
My guess is that these keys aren't in the source data, and that if you traverse through a `&MerkleTreeNode` by reading the raw data that you'll end up in a place where you'll have to have retained information about which data structure you're in in order to read the next bit. Or, is this one of the data structures that you're using for accounting purposes and doesn't relate directly to chain data?

Reply:
Yeah that's correct, these keys don't actually exist in the source data and you'd need to retain information about which data structure you're in in order to read the next bit. Thanks for catching this, I think what we want to do is similar to how this is represented/handled for the various ethereum MMPTs whereby we have this [generic definition](https://github.com/ipld/ipld/blob/master/specs/codecs/dag-eth/state.md#trie-node-ipld) for `MerkleTreeNode`, including the leaf node with this `Value` union, but then we explicitly define specific subtypes of the `MerkleTreeNode` for each type of tree (per value) e.g. how we have [TxTrie](https://github.com/ipld/ipld/blob/master/specs/codecs/dag-eth/state.md#transaction-trie-ipld), [RctTrie](https://github.com/ipld/ipld/blob/master/specs/codecs/dag-eth/state.md#receipt-trie-ipld), etc, for ethereum. And these will each have their own multicodec byte prefix that specifies which value is stored in that particular tree and maps to a codec that can handle the encoding/decoding of the values in that tree.